### DISCUSSION TITLE
AuthApiError: Email link is invalid or has expired

### DISCUSSION THREAD

**Discussion by ryuhangyeong** [supabase/supabase#29432] [Questions] [answered]
```
# Email Template
<h2>Magic Link</h2>

<p>Follow this link to login:</p>
<p><a href="{{ .SiteURL }}/auth/confirm?token_hash={{ .TokenHash }}&type=magiclink">Log In</a></p>
```

```
import { NextResponse } from "next/server";
import { createClient } from "@/utils/supabase/server";

import type { EmailOtpType } from "@supabase/supabase-js";
import type { NextRequest } from "next/server";

export async function GET(request: NextRequest) {
  const { searchParams } = new URL(request.url);
  const token_hash = searchParams.get("token_hash");
  const type = searchParams.get("type") as EmailOtpType | null;
  const next = "/account";

  const redirectTo = request.nextUrl.clone();
  redirectTo.pathname = next;
  redirectTo.searchParams.delete("token_hash");
  redirectTo.searchParams.delete("type");

  if (token_hash && type) {
    const superBaseClient = createClient();

    const { error } = await superBaseClient.auth.verifyOtp({
      type,
      token_hash,
    });

    console.log(error);

    if (!error) {
      redirectTo.searchParams.delete("next");
      return NextResponse.redirect(redirectTo);
    }
  }

  redirectTo.pathname = "/error";
  return NextResponse.redirect(redirectTo);
}
```

<img width="810" alt="image" src="https://github.com/user-attachments/assets/3f752800-dd73-4501-a084-035c3521ae27">

The token is not recognized properly.

I don't know the cause.

Who can help?

**Comment by ryuhangyeong** [comment, marked as answer]
## Self Answer

- `type`

`magiclink` -> `email`

**Reply by kolberszymon** [reply]
I have the same problem, what do you mean by this answer?

**Reply by ryuhangyeong** [reply]
@kolberszymon 

Sorry for the late reply.

I don't know the exact reason, but 

If the type is email, the flow is as follows.

If a specific email address has not yet registered, a membership registration email template will be sent.

If a specific email address is registered, it will be sent as a magic link template.

```
<p><a href="{{ .SiteURL }}/auth/confirm?token_hash={{ .TokenHash }}&type=email">Log In</a></p>
```

Try with the example above.

**Reply by Karsenh** [reply]
I'm facing similar issue and not seeing the answer here.

If an account does NOT exist in supabase and a magic link is sent, it will dispatch a "Confirmation" email. 

Whereas, if the account already exists, it will dispatch a magic link email. 

I've read that `supabase.auth.verityOtp({ type, token_hash}) is the logic to be run on the confirmation email callback route but I'm getting "Email link is invalid or has expired" as a result.

<img width="450" alt="Screenshot 2025-01-28 at 12 49 49 PM" src="https://github.com/user-attachments/assets/ff92f0a2-e1e0-493a-b49a-90c283a9c7db" />

<img width="620" alt="Screenshot 2025-01-28 at 12 50 13 PM" src="https://github.com/user-attachments/assets/e77778cd-66be-46ac-921e-59916f97a7a4" />


I realize I can simply turn off email confirmation in the project settings within supabase but I'd rather be able to handle the email confirmation to use it. How is this possible?

Thank you!

**Reply by Karsenh** [reply]
Nevermind - I just figured out that the type coming through was wrong because I set it wrong in the Confirmation email template within supabase. I was using the magic link template in the confirmation email because I didn't notice the type=magiclink vs type=signup difference.

**Reply by nicholas5538** [reply]
@Karsenh

> Nevermind - I just figured out that the type coming through was wrong because I set it wrong in the Confirmation email template within supabase. I was using the magic link template in the confirmation email because I didn't notice the type=magiclink vs type=signup difference.

Sorry to bother you, I'm currently having this issue only on a production build. There's no issue in the email link whenever I start a new build, but I will start to get this error after a while until I start a fresh build again....

In their documentation, they use type="email" for the confirmation email template. However, it seems to be working now after I switched to `type=signup` and `type=recovery` for confirmation email and reset password respectively, I still need to monitor this though.

Do you have any idea why?
